### DISCUSSION TITLE
Replace matchX options in packageRules with a single "match" string

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#31662] [Suggest an Idea]
### Tell us more.

Now that all "exclude" are gone, and all "match" are AND'd, it's a lot simpler to consolidate these all into a single `match` string which can support `AND`, `OR` and bracketing `()`. The `match` would then behave similarly to SQL `WHERE` statements.

Here's our current list with their matching type:
 - matchBaseBranches: regexOrGlob
 - matchCategories: regexOrGlob
 - matchConfidence: regexOrGlob
 - matchCurrentAge: age string
 - matchCurrentValue: version
 - matchCurrentVersion: single regexOrGlob
 - matchDatasources: regexOrGlob
 - matchDepNames: regexOrGlob
 - matchDepTypes: regexOrGlob
 - matchFileNames: regexOrGlob
 - matchManagers: regexOrGlob
 - matchNewValue: single regexorGlob
 - matchPackageNames: regexOrGlob
 - matchRepositories: regexOrGlob
 - matchSourceUrls: regexOrGlob
 - matchUpdateTypes: regexOrGlob

SQL has the concepts of ANY and ALL plus also NOT.

I think our "regex or glob" concept would be more intuitive/readable if it was split up, e.g. `"matchPackageNames": ["a", "b", "!c", "!d"]` could be `packageName = ANY ["a", "b"] AND packageName != ANY ["c", "d"]`

**Comment by secustor** [comment]
I think as an additional Matcher to cover advanced use cases it makes sense. 

That being said >80% of usecases I see are simple comparisons and for those vaildation and writing of rules would become harder. 
So Im not in favor of replacing the current Matchers.

**Comment by setchy** [comment]
I have a current need where I want to write some package rule which would be dependent/conditional on the presence of another package being used.
